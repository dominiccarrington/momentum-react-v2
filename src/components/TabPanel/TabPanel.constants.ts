import { AllowedTagNames } from './TabPanel.types';

const CLASS_PREFIX = 'md-tab-panel';

const DEFAULTS = {
  as: 'div' as AllowedTagNames,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
