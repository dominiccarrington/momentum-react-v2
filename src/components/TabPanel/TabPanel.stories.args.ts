import { Meta } from '@storybook/react';
import { TabPanelProps } from '.';
import { commonStyles as _commonStyles } from '../../storybook/helper.stories.argtypes';

const tabPanelArgTypes: Meta<TabPanelProps>['argTypes'] = {
  // NOTE: Below is an example. See [Storybook argTypes documentation]{@link https://storybook.js.org/docs/react/api/argtypes}.
  children: {
    description: 'Provides the child nodes for this element.', // NOTE: Description of this prop.
    control: { type: 'text' }, // NOTE: Control type for this prop.
    table: {
      type: {
        summary: 'ReactNode', // NOTE: Explicit type of this prop.
      },
      defaultValue: {
        summary: 'undefined', // NOTE: Default value for this prop.
      },
    },
  },
  tagName: {
    description: 'Define the wrapping element tag',
    control: {
      type: 'select',
      options: ['div', 'section'],
    },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'div',
      },
    },
  },
};

export { tabPanelArgTypes };

const commonStyles = { ..._commonStyles };
delete commonStyles.id;

export default {
  ...commonStyles,
  ...tabPanelArgTypes,
};
