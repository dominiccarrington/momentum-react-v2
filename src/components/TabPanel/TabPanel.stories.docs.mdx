The `<TabPanel />` component implements the a11y requirements for a
[tabpanel](https://w3c.github.io/aria/#tabpanel) based upon the context within a `<Tabs />`
component.
