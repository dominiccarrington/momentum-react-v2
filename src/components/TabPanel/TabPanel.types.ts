import { ComponentProps } from 'react';

export type AllowedTagNames = 'div' | 'section';
export type Props<TComponent extends AllowedTagNames = AllowedTagNames> =
  ComponentProps<TComponent> & {
    as?: TComponent;
  };
