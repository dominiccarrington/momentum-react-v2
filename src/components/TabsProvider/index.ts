import { default as Tabs } from './TabsProvider';
import * as CONSTANTS from './TabsProvider.constants';
import { Props } from './TabsProvider.types';
import { useTabsContext } from './TabsProvider.utils';

export { CONSTANTS as TABS_CONSTANTS };

export type TabsProps = Props;

export default Tabs;
export { useTabsContext };
