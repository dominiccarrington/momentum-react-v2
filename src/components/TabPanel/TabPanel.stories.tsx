import { Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import TabPanel, { TabPanelProps } from './';
import argTypes from './TabPanel.stories.args';
import Documentation from './TabPanel.stories.docs.mdx';

export default {
  title: 'Momentum UI/TabPanel',
  component: TabPanel,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

// NOTE: Primary story. This renders a single component with all external props.
const Example = Template<TabPanelProps>(TabPanel).bind({});

Example.argTypes = { ...argTypes };

Example.args = {
  children: 'Example',
};

export { Example };
