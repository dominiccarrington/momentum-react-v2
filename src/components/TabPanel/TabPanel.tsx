import React, { FC, HTMLAttributes } from 'react';
import classnames from 'classnames';

import { STYLE, DEFAULTS } from './TabPanel.constants';
import { AllowedTagNames, Props } from './TabPanel.types';
import './TabPanel.style.scss';
import { useTabsContext } from '../Tabs/Tabs.utils';

/**
 * The TabPanel Component
 */
export const TabPanel: FC<Props> = <T extends AllowedTagNames>(props: Props<T>) => {
  const { as: Tag = DEFAULTS.as, ...componentProps } = props;

  const tabsContext = useTabsContext();

  if (tabsContext === null) {
    // eslint-disable-next-line no-console
    console.warn('MRV2 (TabPanel): Not recommended to use this component outside of a TabsContext');
  }

  const ariaProps = {
    'aria-labelledby': tabsContext?.activeTabId || props['aria-labelledby'],
    id: tabsContext?.activePanelId,
  } as HTMLAttributes<HTMLElement>;

  return (
    <Tag
      {...(componentProps as HTMLAttributes<HTMLElement>)}
      role="tabpanel"
      className={classnames(STYLE.wrapper, componentProps.className)}
      {...ariaProps}
    />
  );
};

export default TabPanel;
