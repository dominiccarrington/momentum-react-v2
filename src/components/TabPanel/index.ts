import { default as TabPanel } from './TabPanel';
import * as CONSTANTS from './TabPanel.constants';
import { Props } from './TabPanel.types';

export { CONSTANTS as TAB_PANEL_CONSTANTS };

export type TabPanelProps = Props;

export default TabPanel;
