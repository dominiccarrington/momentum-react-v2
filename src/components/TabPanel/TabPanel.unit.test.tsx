import React from 'react';
import { mount } from 'enzyme';

import TabPanel, { TAB_PANEL_CONSTANTS as CONSTANTS } from './';
import TabsProvider from '../TabsProvider';

describe('<TabPanel />', () => {
  let consoleWarnSpy: jest.SpyInstance;

  beforeEach(() => {
    consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {
      /* NOOP */
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<TabPanel>Hello World</TabPanel>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(<TabPanel className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<TabPanel style={style} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with tagName', () => {
      expect.assertions(1);

      const container = mount(<TabPanel tagName="section" />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have the tabpanel role', () => {
      expect.assertions(1);

      const element = mount(<TabPanel />)
        .find(TabPanel)
        .getDOMNode();

      expect(element.getAttribute('role')).toBe('tabpanel');
    });

    it('should have its wrapper class', () => {
      expect.assertions(1);

      const element = mount(<TabPanel />)
        .find(TabPanel)
        .getDOMNode();

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = mount(<TabPanel className={className} />)
        .find(TabPanel)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should not use a provided id', () => {
      expect.assertions(1);

      const id = 'example-id';

      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      const element = mount(<TabPanel id={id} />)
        .find(TabPanel)
        .getDOMNode();

      expect(element.id).not.toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = mount(<TabPanel style={style} />)
        .find(TabPanel)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    it('should render as the provided tag name', () => {
      const container = mount(
        <TabPanel aria-labelledby="label" className="test" tagName="section">
          Hello World
        </TabPanel>
      );

      expect(container.exists()).toBe(true);
      expect(container.find('section').exists()).toBe(true);
      expect(container.find('section').props()).toStrictEqual({
        'aria-labelledby': 'label',
        className: 'md-tab-panel-wrapper test',
        id: undefined,
        children: 'Hello World',
        role: 'tabpanel',
      });
      expect(consoleWarnSpy).toHaveBeenCalled();
    });

    it('should set the id and aria-labelledby based on selectedTab from <TabsProvider />', () => {
      const wrapper = mount(
        <TabsProvider selectedTab={'tab'} id="TabListUtils">
          <TabPanel>Hello World</TabPanel>
        </TabsProvider>
      );

      const container = wrapper.find(TabPanel);
      const element = container.getDOMNode();

      expect(element.getAttribute('id')).toBe('TabListUtilstab-TabPanel');
      expect(element.getAttribute('aria-labelledby')).toBe('TabListUtilstab');
    });
  });

  describe('actions', () => {
    it('renders with a console.warn outside of <TabsProvider />', () => {
      const container = mount(<TabPanel>Hello World</TabPanel>);

      expect(container.exists()).toBe(true);
      expect(consoleWarnSpy).toHaveBeenCalled();
    });
  });
});
