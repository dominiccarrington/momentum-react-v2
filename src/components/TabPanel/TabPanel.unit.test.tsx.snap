// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabPanel /> snapshot should match snapshot 1`] = `
<TabPanel>
  <div
    className="md-tab-panel-wrapper"
    role="tabpanel"
  >
    Hello World
  </div>
</TabPanel>
`;

exports[`<TabPanel /> snapshot should match snapshot with className 1`] = `
<TabPanel
  className="example-class"
>
  <div
    className="md-tab-panel-wrapper example-class"
    role="tabpanel"
  />
</TabPanel>
`;

exports[`<TabPanel /> snapshot should match snapshot with id 1`] = `
<TabPanel
  id="example-id"
>
  <div
    className="md-tab-panel-wrapper"
    id="example-id"
    role="tabpanel"
  />
</TabPanel>
`;

exports[`<TabPanel /> snapshot should match snapshot with style 1`] = `
<TabPanel
  style={
    Object {
      "color": "pink",
    }
  }
>
  <div
    className="md-tab-panel-wrapper"
    role="tabpanel"
    style={
      Object {
        "color": "pink",
      }
    }
  />
</TabPanel>
`;
